src/basic/fs-util.c

```diff
@@ -1028,3 +1028,105 @@ int parse_cifs_service(
 
         return 0;
 }
+
+int resolve_auto_dir(const char *path, bool want_dir, char **ret_path) {
+        _cleanup_free_ char *parent = NULL;
+        _cleanup_free_ char *filename = NULL;
+        _cleanup_free_ char *versioned_path = NULL;
+        _cleanup_closedir_ DIR *dirp = NULL;
+        int r;
+
+        if (!path)
+                return -EINVAL;
+
+        log_debug("Checking for versions of %s.", path);
+
+        const char *auto_dir = path;
+        const char *prefix = NULL;
+        struct stat st;
+        r = stat(path, &st);
+        if (r < 0) {
+                /* Does the path match pattern "*.auto.d/filename"? */
+                r = path_extract_directory(path, &parent);
+                if (r < 0)
+                        return r;
+                r = path_extract_filename(path, &filename);
+                if (r < 0)
+                        return r;
+                auto_dir = parent;
+                prefix = filename;
+                r = stat(auto_dir, &st);
+                if (r < 0)
+                        return r;
```

Comment:
stat() returns errors in errno. thus `return -errno` here. Or alternatively use our new RET_NERRNO() construct